lldb/tools/lldb-server/lldb-platform.cpp

```diff
@@ -421,97 +417,210 @@ int main_platform(int argc, char *argv[]) {
-                          inferior_arguments, log_file, log_channels);
-        if (error.Fail()) {
-          {
-
-            std::lock_guard<std::mutex> guard(gdbserver_portmap_mutex);
-            gdbserver_portmap.FreePort(*available_port);
-          }
-          LLDB_LOGF(GetLog(LLDBLog::Platform), "spawn_process failed: %s",
-                    error.AsCString());
-          WithColor::error()
-              << "spawn_process failed: " << error.AsCString() << "\n";
-        }
-      }
-      // Parent doesn't need a connection to the lldb client
-      delete conn;
-
-      // Parent will continue to listen for new connections.
-      continue;
-    } else {
-      // If not running as a server, this process will not accept
-      // connections while a connection is active.
-      acceptor_up.reset();
-
-      // When not running in server mode, use all available ports
-      platform.SetPortMap(std::move(gdbserver_portmap));
+    llvm::Expected<std::vector<MainLoopBase::ReadHandleUP>> gdb_handles =
+        sock_gdb->Accept(*g_main_loop, [&platform, &gdb_inferior_arguments](
+                                           std::unique_ptr<TCPSocket> sock_up) {
+          printf("Gdb connection established.\n");
+          Status error;
+          Log *log = GetLog(LLDBLog::Platform);
+          SharedSocket shared_socket(sock_up.get(), error);
+          if (error.Success()) {
+            lldb::pid_t child_pid = LLDB_INVALID_PROCESS_ID;
+            std::string socket_name;
+            error = platform.LaunchGDBServer(gdb_inferior_arguments, child_pid,
+                                             socket_name,
+                                             shared_socket.GetSendableFD());
+            if (error.Success() && child_pid != LLDB_INVALID_PROCESS_ID) {
+              error = shared_socket.CompleteSending(child_pid);
+              if (error.Success()) {
+                // Use gdb inferior arguments once.
+                gdb_inferior_arguments.Clear();
+              } else {
+                Host::Kill(child_pid, SIGTERM);
+                LLDB_LOGF(log, "gdbserver CompleteSending failed: %s",
+                          error.AsCString());
+              }
+            }
+          } else
+            LLDB_LOGF(log, "gdbserver SharedSocket failed: %s",
+                      error.AsCString());
+
+          sock_up.reset();
+        });
+    if (!gdb_handles) {
+      printf("Failed to accept gdb: %s\n",
+             llvm::toString(gdb_handles.takeError()).c_str());
+      return socket_error;
     }
 
-    platform.SetConnection(std::unique_ptr<Connection>(conn));
+    do {
+      g_main_loop->Run();
+    } while (!g_terminate);
+
+    gdb_handles->clear();
+    platform_handles->clear();
+
+    g_main_loop.reset();
+    sock_gdb.reset();
+    sock_platform.reset();
+  } else {
+    Socket *conn_up = nullptr;
+    error = sock_named->Accept(conn_up);
```

Comment:
I think it is out of scope of this patch.

Reply:
Unix sockets accept only one socket. A single blocking accept() is enough. I think a MainLoop is redundant for unix sockets. Now the code in platform_named() is clear and consistent. Adding MainLoop and callbacks just will needlessly complicate the code.

Reply:
I don't think so. The added code should be much smaller compared to the number of lines of code in lldb-platform that it will save. The impact is even bigger when you take into acount that lldb-platform is a huge messy monolith whereas the new function will be self-contained and unit-tested.
You don't have to make the existing Accept function go through the MainLoop like I did with TCP. You can just make an independent implementation if that's easier.

I'll also note that lldb-server currently supports accepting multiple connections over domain sockets (--server mode), whereas your implementation does not. if you added support for that, it would be even more duplicated code.

Reply:
You have an idea of ​​what it should look like. I don't. 
I'm not expert in unix sockets. I have no idea how multiple named sockets may be connected simultaneously. 
I don't see how the following code may be reduced using a MainLoop and callbacks
```
  std::unique_ptr<Socket> sock_named = Socket::Create(protocol, false, error);
  error = sock_named->Listen(name, backlog);
  Socket *conn_up = nullptr;
  error = sock_named->Accept(conn_up);
  GDBRemoteCommunicationServerPlatform platform(protocol, 0);
  platform.SetConnection(std::unique_ptr<Connection>(new ConnectionFileDescriptor(conn_up)));
  client_handle(platform, args);
```
Anyway, this shouldn't be a stopper for this patch.

Reply:
It will be reduced (to nothing) if it can use the same mainloop code that used by the tcp path. I think it is a showstopper because it makes the makes the patch very ugly. I can implement the mainloop/accept version for unix sockets, but it will only happen next week.

Reply:
It will reduce the function platform_named() to nothing. But current platform_tcp() and GDBRemoteCommunicationServerPlatform will grow and will become terrible.
Unix sockets do not require ports and accept only 1 connection. GDBRemoteCommunicationServerPlatform has the different logic for TCP and unix sockets. We don't wait for a connection before launching gdbserver to provide fd in case of unix sockets. Using MainLoop for unix sockets will make lldb-platform.cpp and GDBRemoteCommunicationServerPlatform extremely confusing. Currently we don't need a separated child process to handle unix socket connection. Unix sockets are used only locally. The only point in using unix sockets - they are lite. But you want to make the connection process as hard as possible - unnecessary child processes, shared sockets for both lldb-platform and gdbserver, etc. I totally disagree. But if it is the showstopper, I will wait for your implementation of the mainloop/accept version for unix sockets.